File: LeetCode_Java/stack/DailyTemperatures.java
package stack;

import java.util.Arrays;
import java.util.Stack;

public class DailyTemperatures {

    public int[] dailyTemperatures(int[] temperatures) {

        int[] results = new int[temperatures.length];
        Stack<Integer> stack = new Stack<>();

        for (int i = 0; i < temperatures.length; i++) {
            while (!stack.isEmpty() && temperatures[i] > temperatures[stack.peek()]) {
                int index = stack.pop();
                results[index] = i - index;
            }
            stack.push(i);
        }

        return results;

    }

    public static void main(String[] args) {
        DailyTemperatures solution = new DailyTemperatures();
        int[] temperatures = {73, 74, 75, 71, 69, 72, 76, 73};
        int[] result = solution.dailyTemperatures(temperatures);
        System.out.println(Arrays.toString(result));  // Output: [1, 1, 4, 2, 1, 1, 0, 0]
    }
}

File: LeetCode_Python/stack/DailyTemperatures.py
class DailyTemperatures:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        n = len(temperatures)
        res = [0] * n
        stack = []

        for i in range(n):
            while stack and temperatures[i] > temperatures[stack[-1]]:
                index = stack.pop()
                res[index] = i - index
            stack.append(i)

        return res

if __name__ == '__main__':
    solution = DailyTemperatures()
    
    # Test Case 1:
    temps1 = [73, 74, 75, 71, 69, 72, 76, 73]
    print("Test Case 1: Input:", temps1)
    print("Expected Output: [1, 1, 4, 2, 1, 1, 0, 0]")
    print("Actual Output:", solution.dailyTemperatures(temps1))
    print()
    
    # Test Case 2: Strictly decreasing temperatures, no warmer day in future
    temps2 = [80, 79, 78, 77, 76]
    print("Test Case 2: Input:", temps2)
    print("Expected Output: [0, 0, 0, 0, 0]")
    print("Actual Output:", solution.dailyTemperatures(temps2))
    print()
    
    # Test Case 3: Strictly increasing temperatures
    temps3 = [60, 62, 64, 66, 68]
    print("Test Case 3: Input:", temps3)
    print("Expected Output: [1, 1, 1, 1, 0]")
    print("Actual Output:", solution.dailyTemperatures(temps3))
    print()
    
    # Test Case 4: All temperatures are equal
    temps4 = [70, 70, 70, 70]
    print("Test Case 4: Input:", temps4)
    print("Expected Output: [0, 0, 0, 0]")
    print("Actual Output:", solution.dailyTemperatures(temps4))
    print()